File: src - java/Esercizi/CercaFoglia/CercaFoglia.java
	public class RisultatoRicercaFoglia {
		// classe immutabile
		private boolean trovato;
		private boolean foglia;

		public RisultatoRicercaFoglia(boolean trovato,
		        boolean foglia) {
			this.trovato = trovato;
			this.foglia = foglia;
		}

		public boolean getTrovato() {
			return trovato;
		}

		public boolean getFoglia() {
			return foglia;
		}
	}

	RisultatoRicercaFoglia _cercaFoglia(Nodo n, int el) {
		if (n == null) // Se n punta a null
			// uscita anticipata con false
			return new RisultatoRicercaFoglia(false, false);

		RisultatoRicercaFoglia rrf = null;

		// Cerco subito più "in basso".
		if (el > n.elem)
			rrf = _cercaFoglia(n.des, el);
		else
			rrf = _cercaFoglia(n.sin, el);

		if (rrf.trovato) // se in basso l'ho trovato...
			return rrf; // restituisco il verdetto.

		if (n.elem == el) { // se sono proprio io...
			rrf.trovato = true; // ...imposto trovato a true...
			rrf.foglia = eUnaFoglia(n); // ...e verifico se è
			                            // una foglia.
		}

		return rrf;
	}

	public class RisultatoRicercaNodo {
		// classe immutabile
		private boolean trovato;
		private boolean nodo;

		public RisultatoRicercaNodo(boolean trovato, boolean nodo) {
			this.trovato = trovato;
			this.nodo = nodo;
		}

		public boolean getTrovato() {
			return trovato;
		}

		public boolean getNodo() {
			return nodo;
		}
	}

	RisultatoRicercaNodo _cercaNodo(Nodo n, int el) {
		if (n == null)
			return new RisultatoRicercaNodo(false, false);

		if (n.elem == el) { // Se l'elemento corrente è pari ad
							// el...
			RisultatoRicercaNodo rrn = new RisultatoRicercaNodo(
			        false, false);
			rrn.nodo = (!eUnaFoglia(n)); // ...verifico se è un
			                             // nodo...
			rrn.trovato = true; // ...assegno risultato positivo.
			// E' infatti inutile procedere verso il basso.
			return rrn;
		} else // Se non l'ho trovato, cerco più "in basso".
		if (el > n.elem)
			return _cercaNodo(n.des, el);
		else
			return _cercaNodo(n.sin, el);
	}

	// Metodi pubblici
	public RisultatoRicercaFoglia cercaFoglia(int el) {
		return _cercaFoglia(root, el);
	}

	public RisultatoRicercaNodo cercaNodo(int el) {
		return _cercaNodo(root, el);
	}

File: src - c++/Esercizi/CercaFoglia/CercaFoglia.cpp
//Metodi privati
inline bool AlberoBinario::EUnaFoglia(const PNodo& n) const {
  return (!n->des && !n->sin);
}

//Metodi ricorsivi di supporto
bool AlberoBinario::_CercaFoglia(const PNodo& n, TElem el,
                                          bool& foglia) const {
  if (!n)
    return false;
    
  bool trovato;
    
  //Cerco subito più in basso.
  if (el > n->el)
    trovato = _CercaFoglia(n->des, el, foglia);
  else  
    trovato = _CercaFoglia(n->sin, el, foglia);
      
  if (!trovato) { //Se più in basso non l'ho trovato...
    if (n->el == el) { //...e sono proprio io...
      trovato = true; //...imposto trovato a true...
      foglia = EUnaFoglia(n); //...e verifico se è una foglia.
    }    
  }
      
  return trovato;
}

bool AlberoBinario::_CercaNodo(const PNodo& n, TElem el, bool& nodo) const {
  if (!n)
    return false;
    
  if (n->el == el) { //Se l'elemento corrente è pari ad el...
    nodo = (!EUnaFoglia(n)); //...verifico se è un nodo...
    return true; //...ed esco con risultato positivo.
                 //E' infatti inutile procedere verso il basso.
  }
  else //Se non l'ho trovato, cerco più in basso.
    if (el > n->el)
      return _CercaNodo(n->des, el, nodo);
    else  
      return _CercaNodo(n->sin, el, nodo);
}  

// Metodi pubblici
bool AlberoBinario::CercaFoglia(TElem el, bool& foglia) const {
  return _CercaFoglia(root, el, foglia);
}

bool AlberoBinario::CercaNodo(TElem el, bool& nodo) const {
  return _CercaNodo(root, el, nodo);
}
